Assignment2-Ancestors/Q1.py

```
@@ -0,0 +1,44 @@
+from Tree import BinaryTree
+
+def find_ancestors(tree, key):
+    """
+    Find ancestors of the first node with the given value while preorder traversal and save the Nodes in a list that return.
+    :param tree: Tree
+    :param key: any
+    :return: list(Nodes)
+    """
+    node = tree.find(key)
+    if node:

```

Returns `None` if node is not found? Maybe should be mentioned in the docstring above? Also possibly:

```
if not node:
  return None
parent = node.parent
...
```

So that the error handling case would be explicit.

---

I understand you, thanks) 
But still is it better to raise an error when node is not found or should we let the function return None?

---

About last_common_element() function. Yes, there was that silly mistake with indexes and even some others, but funny thing is that results while testing were ok... It was just lucky cases,  or unlucky as I've posted bad solution :)
But I've fixed it and now it is suppose to work good.

---

It can be a lot of work to think of good test cases to catch all reasonable mistakes. I have not directly used for example https://coverage.readthedocs.io/ but I've worked on code bases where something like it had been integrated into workflow to point out test coverage at code review time. That was useful.

I don't know if it would have helped to point out that some code path had no test coverage in this case?